File: November2022/Java/CountCompleteTreeNodes.java
// @saorav21994

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public int countNodes(TreeNode root) {
        
        // < O(n) => O(logn)*O(logn)

        if (root == null)
            return 0;

        TreeNode left, right;
        left = right = root;

        int l = 1, r = 1;      // count of left height ndoes and right height nodes respectively

        while ((left = left.left) != null)
            l += 1;
        while ((right = right.right) != null)
            r += 1;

        if (l == r)
            return (int)Math.pow(2,l) - 1;      // if left and right height is same, then all nodes between them will be fully filled

        return 1 + countNodes(root.left) + countNodes(root.right);

    }
}

File: November2022/C++/count-complete-tree-nodes.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int count(TreeNode*root){
           if(root==NULL){
            return 0;
        }
        if(root->left==NULL && root->right==NULL){
            return 1;
        }
        int lh = 1+ findHeightL(root->left);
        int rh = 1+ findHeightR(root->right);
        
        if(lh==rh){
            // cout<<lh<<" "<<"\n";
            return (1<<lh)-1;
        }
 
        return 1 + count(root->left) + count(root->right);
    }
    int findHeightR(TreeNode*root){
        if(root==NULL){
            return 0;
        }
        return 1 + findHeightR(root->right);
    }
     int findHeightL(TreeNode*root){
        if(root==NULL){
            return 0;
        }
        return 1 + findHeightL(root->left);
    }
public:
    int countNodes(TreeNode* root) {
        return count(root);
    }
};